### PR TITLE
Track Active Storage variants in the database

### PR DESCRIPTION
**Pull request by georgeclaghorn**
Currently, when a variant is requested, we check whether it exists in the service. If it doesn’t, we generate it on-demand and store it for reuse.

The existence check can add unacceptable latency to variant serving. We’ve seen existence checks to third-party services take upwards of 500ms in normal circumstances. Further, checking whether an object exists at a particular key before uploading it [triggers eventual consistency](https://docs.aws.amazon.com/AmazonS3/latest/dev/Introduction.html#ConsistencyModel) in S3:

> Amazon S3 provides read-after-write consistency for PUTS of new objects in your S3 bucket in all Regions with one caveat. The caveat is that if you make a HEAD or GET request to the key name (to find if the object exists) before creating the object, Amazon S3 provides eventual consistency for read-after-write.

This means that when a variant is requested for the first time, generated, and stored in S3, subsequent redirects to the variant’s service URL can fail indefinitely thereafter. The result is sporadically broken images.

This PR addresses the above concerns by doing away with the existence check. When a variant is requested for the first time, we generate it, store it, and track it in the application database. Now we can know whether we’ve already generated a variant without making a remote call to the storage service.

### CODE REVIEW

File: activestorage/app/models/active_storage/variation.rb [hunk 1]
```diff
@@ -58,6 +58,10 @@ def key
     self.class.encode(transformations)
   end
 
+  def digest
+    Digest::SHA1.base64digest Marshal.dump(transformations)
```
Comment (jeremy) [inline activestorage/app/models/active_storage/variation.rb]: Is this consistent across Ruby versions now? Recall some issues with older Ruby where Marshal changes broke e.g. cache keys.

File: activestorage/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
+*   Variants are tracked in the database to avoid existence checks in the storage service.
```
Comment (jeremy) [inline activestorage/CHANGELOG.md]: Let's discuss the upgrade story, too. That you need to migrate your db to get the new table, turn on `track_variants`, and then it'll just work with no backfilling?

File: activestorage/app/models/active_storage/variant_with_record.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+          variation.transform(input, format: "png") do |output|
+            yield io: output, filename: "#{blob.filename.base}.png", content_type: "image/png"
+          end
+        end
+      end
+    end
+
+    def create_or_find_record(image:)
+      @record =
+        ActiveRecord::Base.connected_to(role: ActiveRecord::Base.writing_role) do
```
Comment (jeremy) [inline activestorage/app/models/active_storage/variant_with_record.rb]: May be desirable to extract this method so apps can override to connect to whatever db role. Probably premature.

File: activestorage/app/models/active_storage/variation.rb (hunk 1 above)
Comment (georgeclaghorn) [reply activestorage/app/models/active_storage/variation.rb]: It’s not less consistent than the key we were using before, which indirectly went through `Marshal.dump` (by way of `ActiveSupport::MessageVerifier`).

File: activestorage/CHANGELOG.md (hunk 2 above)
Comment (georgeclaghorn) [reply activestorage/CHANGELOG.md]: I was planning to put that in the upgrading guide when the time came, but I can add it here, too.

File: activestorage/app/models/active_storage/variation.rb (hunk 1 above)
Comment (jeremy) [reply activestorage/app/models/active_storage/variation.rb]: 👍🏼

**Comment by jeremiemv** [comment]
Thanks @georgeclaghorn 
This will be a great improvement :)

**Comment by simonfranzen** [comment]
Hey. This seems like a very nice feature. I am struggling a bit. Is is merged?
I upgraded from rails 5 to rails 6 recently. I have created the active_storage_variant_records table and enabled it in the config, but it's not storing any entries in the db.

**Comment by sedubois** [comment]
@simonfranzen this ships in 6.1
Screenshot 2020-06-08 at 19 26 58

**Comment by buhrmi** [comment]
Can't wait for this to be released. Will there be a way to include variants when querying the database to prevent N+1 queries?

**Comment by vizcay** [comment]
Hi @georgeclaghorn, this is great work and pretty much needed when the S3 server goes down as happened a few times to me and others (#39327).

I've quickly tried to monkey patch ActiveStorage on Rails 6.0 to track already created variants and worked but ended up hitting the problem of the N+1 Query as I wasn't able to preload al the variations when loading all the images.

Before investing more time there, there is no way of using this on Rails 6.0 without bundling all rails from master, right?

**Comment by Cremz** [comment]
This is indeed a great improvement, however the N+1 issue is very frustrating. I rather hit the server and check for the images instead of doing the preloaded query, then 3 x n+1 calls for each variant. (even worse when the model has multiple attached images). Any progress on a fix for this?

**Comment by rept** [comment]
@georgeclaghorn great addition.  Am I right to assume that for existing variants (so already created before this new feature) the check will do done as follows:

- check in DB if variant exists
- if not exists, check in service
- if found there: insert variant record as well
- if not found there: create variant as usual

**Comment by ghiculescu** [comment]
For those concerned about n+1s, please give https://github.com/rails/rails/pull/40842 a try and let me know if it improves things for you.

**Comment by Laykou** [comment]
How does it work with existing variants in db? After deploying this to production will existing variants still work as addressed above? https://github.com/rails/rails/pull/37901#issuecomment-737302497

- check in DB if variant exists
- if not exists, check in service
- if found there: insert variant record as well
- if not found there: create variant as usual

**Comment by louim** [comment]
@Laykou, yes, existing variants will create a new record in the database the first time they are accessed and will skip the service exist call after that.

**Comment by hong-godaddy** [comment]
Are we able to backport this to 5.2?

**Comment by louim** [comment]
Hey @hong-godaddy, features are not backported to earlier versions of Rails.

**Comment by KSH-code** [comment]
S3 supports [Strong consistency model](https://aws.amazon.com/ko/blogs/aws/amazon-s3-update-strong-read-after-write-consistency/) now from Dec. 2020.

Is it possible to change default as false?

**Comment by tbhockey** [comment]
First of all, thank you for this feature 🙏

I'm coming from Rails 5 and have upgrade to 6.1 and have enabled this with `config.active_storage.track_variants = true
`  and while it's working beautifully, but we are getting quite a few broken images. I deleted all `VariantRecord` objects to allow them to rebuild, but strangely this fixed some images and now others are broken. Any idea how to properly resolve this?

**Comment by matthee** [comment]
Hi @tbhockey! I've experienced a similar problem now. Here is how i solved it:

tl;dr run `ActiveStorage::VariantRecord.destroy_all`

First, I tried to delete my variants like so:

```ruby
ActiveStorage::Attachment.where(record_type: "ActiveStorage::VariantRecord").destroy_all
```

This looks fine at first glance. However, ActiveStorage actually creates `ActiveStorage::VariantRecord` records for each variant in the database. When you then browse to a page, where the variant is requested, ActiveStorage looks for the corresponding `ActiveStorage::VariantRecord` record in the database. If it is found, rails figures that the variant must also exist in the storage service (which it does not), leading to a broken image tag in the browser.

In order to fix this problem the `ActiveStorage::VariantRecord` records must also be removed from the database.

Running the following command in the rails console finally fixed my problem and rails happily regenerates all my variants on the fly:

```ruby
ActiveStorage::VariantRecord.destroy_all
```

**Comment by tbhockey** [comment]
Thanks for the reply @matthee . I actually tried this as well, but for me, what happened was after doing that, _different_ images were broken and others were fixed 😆 

We ended up just having to go through and manually delete the Variants for any of the broken images. Since then, everything has been fine. But absolutely no clue what was causing that.
